### ISSUE TITLE
[DataGrid] Roving tabindex is wrong

### ISSUE THREAD

**Issue by oliviertassinari** [mui-org/material-ui-x#1291]
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

Misalignment with the guidelines WAI-ARIA guidelines.

## Expected Behavior 🤔

Alignment with the guidelines.

## Steps to Reproduce 🕹

1. Open: https://codesandbox.io/s/material-demo-forked-c6kb1?file=/demo.js
2. Set the focus on the "start" button
3. Use <kbd>Tab</kbd> to move the focus forward until reaching the "end" button

I see 3 things wrongs:

1. No focus-visible feedback on when the focus is on the root container
2. The root container shouldn't be focusable. See "Because focus is managed using roving tabindex , the grid element is **not focusable**." https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html
3. Per the [roving focus index](https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex) management in a composite widget, setting `tabIndex="0"` on the first item would make more sense. This is the most IMPORTANT part of this bug report. Without it, end-users, can't navigate the cells with the keyboard only.

## Context 🔦

First reported in #1289 by @DanailH.

> as far as I could work out there is no way to navigate to focus a cell in the grid, meaning that when changing focus using tab the focus will move through what is in the toolbar and then be lost - IMO it should focus the first cell or column header (needs to be decided)

Also, looking at the example in https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html. They set tabIndex={-1} on the inner button in the cells. I wonder if we shouldn't do the same with our checkboxes. No idea.

## Your Environment 🌎

v4.0.0-alpha.23

**Comment by dtassone** [comment]
- All tabIndex should be -1
- When you tab into the grid, it should be the first column header to get the focus

It should be fixed as part of #1289

**Comment by dtassone** [comment]
#1253 fixes the cell navigation

**Comment by oliviertassinari** [comment]
Ohh, interesting, I have found a 4th behavior that is not expected.

1. Open https://codesandbox.io/s/material-demo-forked-c6kb1?file=/demo.js
2. Click the first cell
3. Use <kbd>Tab</kbd>

**Actual**: move the focus to the next cell

<img width="232" alt="Screenshot 2021-03-25 at 17 51 07" src="https://user-images.githubusercontent.com/3165635/112511440-b668df80-8d92-11eb-9619-4deb0d97e7e7.png">

**Expected**: move the focus of the pagination previous page icon.

<img width="146" alt="Screenshot 2021-03-25 at 17 51 15" src="https://user-images.githubusercontent.com/3165635/112511443-b7017600-8d92-11eb-83c0-3c4346ae30e6.png">

**Comment by dtassone** [comment]
Ag grid move to the next cell on tab

**Comment by dtassone** [comment]
Actually thinking about it again.

Tab should always go to the next button. So if the column Header has a sortable column or a column menu then it should be tabable. If a cell is editable or has a button it should be tabable. But if a cell is readonly and there is absolutely no button or no action in it then it should be skipped when pressing tab, same for headers and so on...

**Comment by oliviertassinari** [comment]
> Tab should always go to the next button.

@dtassone Why? It's contradictory to what's encouraged in https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex.

**Comment by dtassone** [comment]
Well not really!

The source of truth is https://www.w3.org/TR/wai-aria-practices/#grid

In a grid, every cell contains a focusable element or is itself focusable, regardless of whether the cell content is editable or interactive. **There is one exception: if column or row header cells do not provide functions, such as sort or filter, they do not need to be focusable**. One reason it is important for all cells to be able to receive or contain keyboard focus is that screen readers will typically be in their application reading mode, rather than their document reading mode, when users are interacting with the grid. While in application mode, a screen reader user hears only focusable elements and content that labels focusable elements. So, screen reader users may unknowingly overlook elements contained in a grid that are either not focusable or not used to label a column or row.

Tab: moves focus to the next widget in the grid. Optionally, the focus movement may wrap inside a single cell or within the grid itself.

**Comment by oliviertassinari** [comment]
@dtassone Thanks for linking the source, I would add these important precisions:

> In a grid, every cell contains a focusable element or is itse […]

focusable is strictly different from tabbable.

> Tab: moves focus to the next widget in the grid. Optionally, the focus movement may wrap inside a single cell or within the grid itself.

Is inside the following heading structure:

- h3: 3.12 Grids : Interactive Tabular Data and Layout Containers
- h4: Keyboard Interaction - Setting Focus and Navigating Inside Cells
- h5: Editing and Navigating Inside a Cell

It's about what happens with <kbd>Tab</kbd> **once** you have the focus inside a cell.

**Comment by DanailH** [comment]
@oliviertassinari I think in the process all 3 issues were fixed so it seems that this can be closed. I'll double-check again.

**Comment by oliviertassinari** [comment]
@DanailH Regarding closing, I would propose we do it once we get:

- [ ] A end-to-end test with <kbd>Tab</kbd>
- [ ] `tabIndex={-1}` on the select all checkbox

I can't spot anything else. Some might argue that the data grid should also take care of adding tabIndex={-1} on the child of the cells but seems quite opinionated and slow to perform. I would vote for trying to get away without.

**Comment by DanailH** [comment]
Alright, yup I agree, we can include those 2 points in the ticket DOD.

**Comment by DanailH** [comment]
I think we can close this one. The two points raised above are fixed / added.

**Comment by oliviertassinari** [comment]
@DanailH We only miss a test case, I believe.

**Comment by oliviertassinari** [comment]
There might also be a visual issue, see:

<img width="158" alt="Screenshot 2021-05-05 at 12 38 03" src="https://user-images.githubusercontent.com/3165635/117129270-c49d1900-ad9e-11eb-8f45-eeffbb6ae5aa.png">
<img width="139" alt="Screenshot 2021-05-05 at 12 38 12" src="https://user-images.githubusercontent.com/3165635/117129273-c535af80-ad9e-11eb-8476-b5c54fa88324.png">
<img width="144" alt="Screenshot 2021-05-05 at 12 38 08" src="https://user-images.githubusercontent.com/3165635/117129271-c535af80-ad9e-11eb-8c75-db2967dc5ba5.png">

**Comment by DanailH** [comment]
@oliviertassinari there is this one https://github.com/mui-org/material-ui-x/blob/master/test/e2e/index.test.ts#L76?

I will check the visual regression.
